File: javascript/util/node/web_socket/draft76_parser.js
Faye.WebSocket.Draft76Parser = Faye.Class(Faye.WebSocket.Draft75Parser, {
  version: 'draft-76',
  
  handshakeResponse: function(head) {
    var request = this._socket.request,
        stream  = this._stream,
        
        key1    = request.headers['sec-websocket-key1'],
        value1  = this._numberFromKey(key1) / this._spacesInKey(key1),
        
        key2    = request.headers['sec-websocket-key2'],
        value2  = this._numberFromKey(key2) / this._spacesInKey(key2),
        
        MD5     = crypto.createHash('md5');
    
    MD5.update(this._bigEndian(value1));
    MD5.update(this._bigEndian(value2));
    MD5.update(head.toString('binary'));
    
    try {
      stream.write('HTTP/1.1 101 Web Socket Protocol Handshake\r\n', 'binary');
      stream.write('Upgrade: WebSocket\r\n', 'binary');
      stream.write('Connection: Upgrade\r\n', 'binary');
      stream.write('Sec-WebSocket-Origin: ' + request.headers.origin + '\r\n', 'binary');
      stream.write('Sec-WebSocket-Location: ' + this._socket.url + '\r\n\r\n', 'binary');
      stream.write(MD5.digest('binary'), 'binary');
    } catch (e) {
      // socket closed while writing
      // no handshake sent; client will stop using WebSocket
    }
  },
  
  _numberFromKey: function(key) {
    return parseInt(key.match(/[0-9]/g).join(''), 10);
  },
  
  _spacesInKey: function(key) {
    return key.match(/ /g).length;
  },
  
  _bigEndian: function(number) {
    var string = '';
    Faye.each([24,16,8,0], function(offset) {
      string += String.fromCharCode(number >> offset & 0xFF);
    });
    return string;
  }  
});

File: lib/faye/util/web_socket/draft76_parser.rb
require 'digest/md5'

module Faye
  class WebSocket
    
    class Draft76Parser < Draft75Parser
      def version
        'draft-76'
      end
      
      def handshake_response
        env = @socket.env
        
        key1   = env['HTTP_SEC_WEBSOCKET_KEY1']
        value1 = number_from_key(key1) / spaces_in_key(key1)
        
        key2   = env['HTTP_SEC_WEBSOCKET_KEY2']
        value2 = number_from_key(key2) / spaces_in_key(key2)
        
        hash = Digest::MD5.digest(big_endian(value1) +
                                  big_endian(value2) +
                                  env['rack.input'].read)
        
        upgrade =  "HTTP/1.1 101 Web Socket Protocol Handshake\r\n"
        upgrade << "Upgrade: WebSocket\r\n"
        upgrade << "Connection: Upgrade\r\n"
        upgrade << "Sec-WebSocket-Origin: #{env['HTTP_ORIGIN']}\r\n"
        upgrade << "Sec-WebSocket-Location: #{@socket.url}\r\n"
        upgrade << "\r\n"
        upgrade << hash
        upgrade
      end
      
    private
      
      def number_from_key(key)
        key.scan(/[0-9]/).join('').to_i(10)
      end
      
      def spaces_in_key(key)
        key.scan(/ /).size
      end
      
      def big_endian(number)
        string = ''
        [24,16,8,0].each do |offset|
          string << (number >> offset & 0xFF).chr
        end
        string
      end
    end
    
  end
end

